skbio/diversity/_util.py

```diff
@@ -37,27 +38,39 @@ def _validate_counts_matrix(counts, ids=None, suppress_cast=False):
     # handle case of where counts is a single vector by making it a matrix.
     # this has to be done before forcing counts into an ndarray because we
     # don't yet know that all of the entries are of equal length
-    if len(counts) == 0 or not isinstance(counts[0], collections.Iterable):
-        counts = [counts]
-    counts = np.asarray(counts)
-    if counts.ndim > 2:
-        raise ValueError("Only 1-D and 2-D array-like objects can be provided "
-                         "as input. Provided object has %d dimensions." %
-                         counts.ndim)
-
-    if ids is not None and len(counts) != len(ids):
-        raise ValueError(
-            "Number of rows in ``counts`` must be equal to number of provided "
-            "``ids``.")
-
-    lens = []
-    for v in counts:
-        results.append(_validate_counts_vector(v, suppress_cast))
-        lens.append(len(v))
-    if len(set(lens)) > 1:
-        raise ValueError("All rows in ``counts`` must be of equal length.")
-
-    return np.asarray(results)
+    if isinstance(counts, pd.core.frame.DataFrame):
+        if ids is not None and len(counts.index) != len(ids):
+            raise ValueError(
+                "Number of rows in ``counts``"
+                "must be equal to number of provided ``ids``."
```

Comment:
missing space before must

Reply:
- [x] Unit tests for instantiation with a DataFrame.
- [x] Description of these changes in the change log.
- [x] Expand on the documentation for the DistanceMatrix and DisimilarityMatrix objects. The docs should note that dataframes are supported to instantiate the class.
 - *Also, it would be good if you could check the object is 2D and of a square form (unless that already happens elsewhere - I am not 100% familiar with this code)*.  -> The output of `skbio.diversity.beta.pw_distances()` should be a square matrix, not necessarily the input
- *Any thoughts of whether or not we can take the index and column labels and use those as the matrix ids? This would be handy and it would be a nice sanity check.*  -> I vote for keeping the API identical for the moment.